pkg/controller/helpers.go

```diff
@@ -174,3 +177,95 @@ func deleteKeyFromJSON(jsonWithEtag []byte, keysToDelete ...string) ([]byte, err
 	}
 	return json.Marshal(m)
 }
+
+// convert a list of backendaddress to ip addresses
+func (c *AppGwIngressController) getIPAddresses(addressPools *[]n.ApplicationGatewayBackendAddress) []agpoolv1beta1.BackendAddress {
+	if addressPools == nil {
+		return []agpoolv1beta1.BackendAddress{}
+	}
+
+	var ipAddresses []agpoolv1beta1.BackendAddress
+	for _, addressPool := range *addressPools {
+		address := agpoolv1beta1.BackendAddress{IPAddress: *addressPool.IPAddress}
+		ipAddresses = append(ipAddresses, address)
+	}
+	return ipAddresses
+}
+
+// compare generated BackendAddressPools with cached one
+func (c *AppGwIngressController) isBackendAddressPoolsUpdated(generated, existing *[]n.ApplicationGatewayBackendAddressPool) bool {
+	if existing == nil && generated == nil {
+		return false
+	}
+
+	if existing == nil || generated == nil {
+		return true
+	}
+
+	backendIDtoIPAddressesMapNew := make(map[string][]string)
+	backendIDtoIPAddressesMapExisting := make(map[string]bool)
+
+	for _, gbap := range *generated {
+		backendIDNew := gbap.ID
+		glog.V(9).Infof("New: find backend pool id: [%s]", *backendIDNew)
+		ips := make([]string, len(*gbap.BackendAddresses))
+		for i, ip := range *gbap.BackendAddresses {
+			ips[i] = *ip.IPAddress
+		}
+		sort.Strings(ips)
+		backendIDtoIPAddressesMapNew[*backendIDNew] = ips
+	}
+
+	for _, cbap := range *existing {
+		backendIDExisting := cbap.ID
+		backendIDtoIPAddressesMapExisting[*backendIDExisting] = true
+		if ipAddresses, exists := backendIDtoIPAddressesMapNew[*backendIDExisting]; exists {
+			glog.V(9).Infof("Existing: find backend pool name: %s", *backendIDExisting)
+
+			ips := make([]string, len(*cbap.BackendAddresses))
+			for i, ip := range *cbap.BackendAddresses {
+				ips[i] = *ip.IPAddress
+			}
+			sort.Strings(ips)
+			// indicates the backend pool is updated
+			if !equal(ipAddresses, ips) {
+				glog.V(5).Infof("New: backend address pool ip: %v", ipAddresses)
+				glog.V(5).Infof("Existing: backend address pool ip: %v", ips)
+				return true
+			}
+		} else {
+			// if any backend pool is deleted
+			glog.V(3).Infof("Existing: NOT find backend pool id: [%s]", *backendIDExisting)
+			return true
+		}
+	}
+
+	// if any backend pool is newly added
+	for _, gbap := range *generated {
```

Comment:
we can remove `backendIDtoIPAddressesMapExisting` and this for loop. This will be covered in the length check suggested above.